**This notebook is an exercise in the [Introduction to Machine Learning](https://www.kaggle.com/learn/intro-to-machine-learning) course.  You can reference the tutorial at [this link](https://www.kaggle.com/alexisbcook/machine-learning-competitions).**

---

# Introduction

In this exercise, you will create and submit predictions for a Kaggle competition. You can then improve your model (e.g. by adding features) to apply what you've learned and move up the leaderboard.

Begin by running the code cell below to set up code checking and the filepaths for the dataset.

In [1]:
# Set up code checking
from learntools.core import binder
binder.bind(globals())
from learntools.machine_learning.ex7 import *

# Set up filepaths
import os
if not os.path.exists("../input/train.csv"):
    os.symlink("../input/home-data-for-ml-course/train.csv", "../input/train.csv")  
    os.symlink("../input/home-data-for-ml-course/test.csv", "../input/test.csv")

Here's some of the code you've written so far. Start by running it again.

In [2]:
# Import helpful libraries
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import train_test_split

# Load the data, and separate the target
iowa_file_path = '../input/train.csv'
home_data = pd.read_csv(iowa_file_path)
y = home_data.SalePrice

# Create X (After completing the exercise, you can return to modify this line!)
features = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']

# Select columns corresponding to features, and preview the data
X = home_data[features]
X.head()

# Split into validation and training data
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state=1)

# Define a random forest model
rf_model = RandomForestRegressor(random_state=1)
rf_model.fit(train_X, train_y)
rf_val_predictions = rf_model.predict(val_X)
rf_val_mae = mean_absolute_error(rf_val_predictions, val_y)

print("Validation MAE for DEFAULT Random Forest Model: {:,.0f}".format(rf_val_mae))

Validation MAE for DEFAULT Random Forest Model: 21,857


# Train a model for the competition

The code cell above trains a Random Forest model on **`train_X`** and **`train_y`**.  

Use the code cell below to build a Random Forest model and train it on all of **`X`** and **`y`**.

In [3]:
# To improve accuracy, create a new Random Forest model which you will train on all training data
rf_model_on_full_data = RandomForestRegressor(random_state=1)

# fit rf_model_on_full_data on all data from the training data
rf_model_on_full_data.fit(X, y)

RandomForestRegressor(random_state=1)

Now, read the file of "test" data, and apply your model to make predictions.

In [4]:
# path to file you will use for predictions
test_data_path = '../input/test.csv'

# read test data file using pandas
test_data = pd.read_csv(test_data_path)

# create test_X which comes from test_data but includes only the columns you used for prediction.
# The list of columns is stored in a variable called features
test_X = test_data[features]

# make predictions which we will submit. 
test_preds_defaultmodel = rf_model_on_full_data.predict(test_X)

test_preds_defaultmodel[0:15].round(0).tolist()

[122657.0,
 156789.0,
 182959.0,
 178102.0,
 189049.0,
 180979.0,
 172797.0,
 173717.0,
 187535.0,
 116172.0,
 190798.0,
 93823.0,
 89249.0,
 145111.0,
 124696.0]

Before submitting, run a check to make sure your `test_preds` have the right format.

# Results default model

First 15 predictions:

[122657.0,
 156789.0,
 182959.0,
 178102.0,
 189049.0,
 180979.0,
 172797.0,
 173717.0,
 187535.0,
 116172.0,
 190798.0,
 93823.0,
 89249.0,
 145111.0,
 124696.0]
 
Validation MAE for DEFAULT Random Forest Model: **21,857**



# Model optimization

· Will clean the dataset, hot-code categorical features, scale numerical features, and train a new model

## Cleaning dataset

In [5]:
import pandas as pd
import numpy as np

In [6]:
pd.set_option('display.max_rows', None)

In [7]:
home_data2 = home_data.copy()
home_data2.set_index('Id', inplace=True)

In [8]:
#Drop columns
threshold = 200

for col in home_data2.columns:
    if home_data2[col].isnull().sum() >= threshold:
        del(home_data2[col])

In [9]:
#Inspect remaining columns with NaNs
columns_with_nan = []

for col in home_data2.columns:
    if home_data2[col].isnull().sum() > 0:
        columns_with_nan.append(col)

home_data2[columns_with_nan].head(10)  
        

MasVnrArea BsmtQual BsmtCond BsmtExposure BsmtFinType1 BsmtFinType2  \
Id                                                                        
1        196.0       Gd       TA           No          GLQ          Unf   
2          0.0       Gd       TA           Gd          ALQ          Unf   
3        162.0       Gd       TA           Mn          GLQ          Unf   
4          0.0       TA       Gd           No          ALQ          Unf   
5        350.0       Gd       TA           Av          GLQ          Unf   
6          0.0       Gd       TA           No          GLQ          Unf   
7        186.0       Ex       TA           Av          GLQ          Unf   
8        240.0       Gd       TA           Mn          ALQ          BLQ   
9          0.0       TA       TA           No          Unf          Unf   
10         0.0       TA       TA           No          GLQ          Unf   

   Electrical GarageType  GarageYrBlt GarageFinish GarageQual GarageCond  
Id                                                                        
1       SBrkr     Attchd       2003.0          RFn         TA         TA  
2       SBrkr     Attchd       1976.0          RFn         TA         TA  
3       SBrkr     Attchd       2001.0          RFn         TA         TA  
4       SBrkr     Detchd       1998.0          Unf         TA         TA  
5       SBrkr     Attchd       2000.0          RFn         TA         TA  
6       SBrkr     Attchd       1993.0          Unf         TA         TA  
7       SBrkr     Attchd       2004.0          RFn         TA         TA  
8       SBrkr     Attchd       1973.0          RFn         TA         TA  
9       FuseF     Detchd       1931.0          Unf         Fa         TA  
10      SBrkr     Attchd       1939.0          RFn         Gd         TA

In [10]:
home_data3 = home_data2.copy()

#Fill NaNs with mean for numerical and mode for rest
columns_with_nan = []

for col in home_data3.columns:
    if home_data3[col].isnull().sum() > 0 and home_data3[col].dtype == 'float64':
        home_data3[col] = home_data3[col].fillna(value = home_data3[col].mean())
    elif home_data3[col].isnull().sum() > 0 and home_data3[col].dtype == 'object':
        home_data3[col] = home_data3[col].fillna(value = "non-existent")
        

In [11]:
home_data3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1460 entries, 1 to 1460
Data columns (total 73 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   MSSubClass     1460 non-null   int64  
 1   MSZoning       1460 non-null   object 
 2   LotArea        1460 non-null   int64  
 3   Street         1460 non-null   object 
 4   LotShape       1460 non-null   object 
 5   LandContour    1460 non-null   object 
 6   Utilities      1460 non-null   object 
 7   LotConfig      1460 non-null   object 
 8   LandSlope      1460 non-null   object 
 9   Neighborhood   1460 non-null   object 
 10  Condition1     1460 non-null   object 
 11  Condition2     1460 non-null   object 
 12  BldgType       1460 non-null   object 
 13  HouseStyle     1460 non-null   object 
 14  OverallQual    1460 non-null   int64  
 15  OverallCond    1460 non-null   int64  
 16  YearBuilt      1460 non-null   int64  
 17  YearRemodAdd   1460 non-null   int64  
 18  RoofStyle    

## Hot-code categorical features and scale numerical features

In [12]:
home_data4 = home_data3.copy()

In [13]:
"""
#Minmax scale of numeric variables
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

# create list of numerical variables
l_num = []
target = 'SalePrice'
for col in home_data4.columns:
    if col != target:
        if home_data4[col].dtype.kind == 'i' or home_data4[col].dtype.kind == 'f':
            l_num.append(col)
        
# scale numerical variables
home_data4[l_num] = scaler.fit_transform(home_data4[l_num])"""

"\n#Minmax scale of numeric variables\nfrom sklearn.preprocessing import MinMaxScaler\nscaler = MinMaxScaler()\n\n# create list of numerical variables\nl_num = []\ntarget = 'SalePrice'\nfor col in home_data4.columns:\n    if col != target:\n        if home_data4[col].dtype.kind == 'i' or home_data4[col].dtype.kind == 'f':\n            l_num.append(col)\n        \n# scale numerical variables\nhome_data4[l_num] = scaler.fit_transform(home_data4[l_num])"

In [14]:
# create list of categorical variables
l_cat = []
for col in home_data4.columns:
    if home_data4[col].dtype.kind == 'O':
        l_cat.append(col)

# create dummy variables
home_data_dummies = pd.get_dummies(home_data4, columns=l_cat, dtype='int32')

## Correlation and elimination of features

In [15]:
corr_matrix = home_data_dummies.corr()

# Transform the upper-right triangle to zeros
upper_triangle_mask = np.triu(np.ones(corr_matrix.shape), k=1).astype(bool)
corr_matrix_upper_zero = corr_matrix.mask(upper_triangle_mask)

# Print the pairs of correlated variables to SalePrice
corr_result = corr_matrix_upper_zero.unstack().sort_values(ascending=False)
corr_result_df = corr_result.reset_index()

In [16]:
print(corr_result_df[corr_result_df['level_0'] != corr_result_df['level_1']].head(50).to_string(index=False))

                  level_0                   level_1        0
       Exterior1st_CBlock        Exterior2nd_CBlock 1.000000
  GarageType_non-existent GarageFinish_non-existent 1.000000
  GarageType_non-existent   GarageQual_non-existent 1.000000
  GarageType_non-existent   GarageCond_non-existent 1.000000
GarageFinish_non-existent   GarageQual_non-existent 1.000000
  GarageQual_non-existent   GarageCond_non-existent 1.000000
GarageFinish_non-existent   GarageCond_non-existent 1.000000
    BsmtQual_non-existent     BsmtCond_non-existent 1.000000
    BsmtCond_non-existent BsmtFinType1_non-existent 1.000000
    BsmtQual_non-existent BsmtFinType1_non-existent 1.000000
             SaleType_New     SaleCondition_Partial 0.986819
    BsmtCond_non-existent BsmtFinType2_non-existent 0.986408
    BsmtQual_non-existent BsmtFinType2_non-existent 0.986408
BsmtFinType1_non-existent BsmtFinType2_non-existent 0.986408
BsmtExposure_non-existent BsmtFinType1_non-existent 0.986408
    BsmtCond_non-existen

In [17]:
corr_result_df[corr_result_df[0] < 0].head()

level_0                  level_1             0
13879      MSZoning_C (all)             ExterCond_Gd -3.520489e-18
13880  Neighborhood_Edwards             ExterCond_Gd -1.580520e-17
13881          ExterCond_Gd             PavedDrive_N -1.750494e-17
13882           GarageYrBlt  GarageQual_non-existent -1.409657e-15
13883           GarageYrBlt  GarageType_non-existent -1.409657e-15

In [18]:
# Get features with np.abs(corr to price) > 0.4

corr_result_df.columns = ['feature_1', 'feature_2', 'corr']
high_corr_features = corr_result_df[(np.abs(corr_result_df['corr']) > 0.4) & (np.abs(corr_result_df['corr']) < 0.8)]

high_corr_features_unique = set(high_corr_features['feature_1'].tolist() + high_corr_features['feature_2'].tolist())

In [19]:
len(high_corr_features_unique)

156

In [20]:
high_corr_features_unique.remove('SalePrice')

## Optimizing model

### Cleaning test data 

In [21]:
test_data = pd.read_csv(test_data_path)
test_data.set_index('Id', inplace=True)

In [22]:
test_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1459 entries, 1461 to 2919
Data columns (total 79 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   MSSubClass     1459 non-null   int64  
 1   MSZoning       1455 non-null   object 
 2   LotFrontage    1232 non-null   float64
 3   LotArea        1459 non-null   int64  
 4   Street         1459 non-null   object 
 5   Alley          107 non-null    object 
 6   LotShape       1459 non-null   object 
 7   LandContour    1459 non-null   object 
 8   Utilities      1457 non-null   object 
 9   LotConfig      1459 non-null   object 
 10  LandSlope      1459 non-null   object 
 11  Neighborhood   1459 non-null   object 
 12  Condition1     1459 non-null   object 
 13  Condition2     1459 non-null   object 
 14  BldgType       1459 non-null   object 
 15  HouseStyle     1459 non-null   object 
 16  OverallQual    1459 non-null   int64  
 17  OverallCond    1459 non-null   int64  
 18  YearBuilt 

In [23]:
#Drop columns
threshold = 200

for col in home_data.columns:
    if home_data[col].isnull().sum() >= threshold:
        del(test_data[col])

In [24]:
test_data2 = test_data.copy()

#Inspect remaining columns with NaNs
columns_with_nan = []

for col in test_data2.columns:
    if test_data2[col].isnull().sum() > 0:
        columns_with_nan.append(col)

test_data2[columns_with_nan].head(10)  

MSZoning Utilities Exterior1st Exterior2nd  MasVnrArea BsmtQual BsmtCond  \
Id                                                                              
1461       RH    AllPub     VinylSd     VinylSd         0.0       TA       TA   
1462       RL    AllPub     Wd Sdng     Wd Sdng       108.0       TA       TA   
1463       RL    AllPub     VinylSd     VinylSd         0.0       Gd       TA   
1464       RL    AllPub     VinylSd     VinylSd        20.0       TA       TA   
1465       RL    AllPub     HdBoard     HdBoard         0.0       Gd       TA   
1466       RL    AllPub     HdBoard     HdBoard         0.0       Gd       TA   
1467       RL    AllPub     HdBoard     HdBoard         0.0       Gd       TA   
1468       RL    AllPub     VinylSd     VinylSd         0.0       Gd       TA   
1469       RL    AllPub     HdBoard     HdBoard         0.0       Gd       TA   
1470       RL    AllPub     Plywood     Plywood         0.0       TA       TA   

     BsmtExposure BsmtFinType1  BsmtFinSF1  ... KitchenQual  Functional  \
Id                                          ...                           
1461           No          Rec       468.0  ...          TA         Typ   
1462           No          ALQ       923.0  ...          Gd         Typ   
1463           No          GLQ       791.0  ...          TA         Typ   
1464           No          GLQ       602.0  ...          Gd         Typ   
1465           No          ALQ       263.0  ...          Gd         Typ   
1466           No          Unf         0.0  ...          TA         Typ   
1467           No          ALQ       935.0  ...          TA         Typ   
1468           No          Unf         0.0  ...          TA         Typ   
1469           Gd          GLQ       637.0  ...          Gd         Typ   
1470           No          ALQ       804.0  ...          TA         Typ   

      GarageType  GarageYrBlt  GarageFinish  GarageCars GarageArea GarageQual  \
Id                                                                              
1461      Attchd       1961.0           Unf         1.0      730.0         TA   
1462      Attchd       1958.0           Unf         1.0      312.0         TA   
1463      Attchd       1997.0           Fin         2.0      482.0         TA   
1464      Attchd       1998.0           Fin         2.0      470.0         TA   
1465      Attchd       1992.0           RFn         2.0      506.0         TA   
1466      Attchd       1993.0           Fin         2.0      440.0         TA   
1467      Attchd       1992.0           Fin         2.0      420.0         TA   
1468      Attchd       1998.0           Fin         2.0      393.0         TA   
1469      Attchd       1990.0           Unf         2.0      506.0         TA   
1470      Attchd       1970.0           Fin         2.0      525.0         TA   

     GarageCond  SaleType  
Id                         
1461         TA        WD  
1462         TA        WD  
1463         TA        WD  
1464         TA        WD  
1465         TA        WD  
1466         TA        WD  
1467         TA        WD  
1468         TA        WD  
1469         TA        WD  
1470         TA        WD  

[10 rows x 26 columns]

In [25]:
test_data3 = test_data2.copy()

#Fill NaNs with mean for numerical and mode for rest
columns_with_nan = []

for col in test_data3.columns:
    if test_data3[col].isnull().sum() > 0 and test_data3[col].dtype == 'float64':
        test_data3[col] = test_data3[col].fillna(value = test_data3[col].mean())
    elif test_data3[col].isnull().sum() > 0 and test_data3[col].dtype == 'object':
        test_data3[col] = test_data3[col].fillna(value = "non-existent")

In [26]:
"""test_data4 = test_data3.copy()
#Minmax scale of numeric variables
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

l_num = []
for col in test_data4.columns:
    if test_data4[col].dtype.kind == 'i' or test_data4[col].dtype.kind == 'f':
        l_num.append(col)
        
# scale numerical variables
test_data4[l_num] = scaler.fit_transform(test_data4[l_num])"""


"test_data4 = test_data3.copy()\n#Minmax scale of numeric variables\nfrom sklearn.preprocessing import MinMaxScaler\nscaler = MinMaxScaler()\n\nl_num = []\nfor col in test_data4.columns:\n    if test_data4[col].dtype.kind == 'i' or test_data4[col].dtype.kind == 'f':\n        l_num.append(col)\n        \n# scale numerical variables\ntest_data4[l_num] = scaler.fit_transform(test_data4[l_num])"

In [27]:
test_data4 = test_data3.copy()

In [28]:
# create list of categorical variables
l_cat = []
for col in test_data4.columns:
    if test_data4[col].dtype.kind == 'O':
        l_cat.append(col)

# create dummy variables
test_data_dummies = pd.get_dummies(test_data4, columns=l_cat, dtype='int32')


In [29]:
test_data_dummies.head()

MSSubClass  LotArea  OverallQual  OverallCond  YearBuilt  YearRemodAdd  \
Id                                                                             
1461          20    11622            5            6       1961          1961   
1462          20    14267            6            6       1958          1958   
1463          60    13830            5            5       1997          1998   
1464          60     9978            6            6       1998          1998   
1465         120     5005            8            5       1992          1992   

      MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  ...  SaleType_New  \
Id                                                   ...                 
1461         0.0       468.0       144.0      270.0  ...             0   
1462       108.0       923.0         0.0      406.0  ...             0   
1463         0.0       791.0         0.0      137.0  ...             0   
1464        20.0       602.0         0.0      324.0  ...             0   
1465         0.0       263.0         0.0     1017.0  ...             0   

      SaleType_Oth  SaleType_WD  SaleType_non-existent  SaleCondition_Abnorml  \
Id                                                                              
1461             0            1                      0                      0   
1462             0            1                      0                      0   
1463             0            1                      0                      0   
1464             0            1                      0                      0   
1465             0            1                      0                      0   

      SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
Id                                                                        
1461                      0                     0                     0   
1462                      0                     0                     0   
1463                      0                     0                     0   
1464                      0                     0                     0   
1465                      0                     0                     0   

      SaleCondition_Normal  SaleCondition_Partial  
Id                                                 
1461                     1                      0  
1462                     1                      0  
1463                     1                      0  
1464                     1                      0  
1465                     1                      0  

[5 rows x 265 columns]

# Train a new model

In [30]:
import random
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error

In [31]:
# Function to optimize max_leaf_nodes
def get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y):
    model = DecisionTreeRegressor(max_leaf_nodes=max_leaf_nodes, random_state=0)
    model.fit(train_X, train_y)
    preds_val = model.predict(val_X)
    mae = mean_absolute_error(val_y, preds_val)
    return(mae)

In [32]:
"""
X = home_data_dummies[list(high_corr_features_unique)]
y = home_data_dummies.SalePrice

train_X, val_X, train_y, val_y = train_test_split(X, y, random_state=1)

for max_leaf_nodes in range(10, 10000, 50):
    rf_model = RandomForestRegressor(random_state=1, max_leaf_nodes=max_leaf_nodes)
    rf_model.fit(train_X, train_y)
    rf_val_predictions = rf_model.predict(val_X)
    rf_val_mae = mean_absolute_error(rf_val_predictions, val_y)
    
    print("Max leaf nodes: %d \t\t Mean Absolute Error: %d" %(max_leaf_nodes, rf_val_mae))
    
    if rf_val_mae < 15000:
        print("Stopping criteria reached.")
        break
"""

'\nX = home_data_dummies[list(high_corr_features_unique)]\ny = home_data_dummies.SalePrice\n\ntrain_X, val_X, train_y, val_y = train_test_split(X, y, random_state=1)\n\nfor max_leaf_nodes in range(10, 10000, 50):\n    rf_model = RandomForestRegressor(random_state=1, max_leaf_nodes=max_leaf_nodes)\n    rf_model.fit(train_X, train_y)\n    rf_val_predictions = rf_model.predict(val_X)\n    rf_val_mae = mean_absolute_error(rf_val_predictions, val_y)\n    \n    print("Max leaf nodes: %d \t\t Mean Absolute Error: %d" %(max_leaf_nodes, rf_val_mae))\n    \n    if rf_val_mae < 15000:\n        print("Stopping criteria reached.")\n        break\n'

In [33]:
"""X = home_data_dummies[list(high_corr_features_unique)]
#X = home_data_dummies.drop(['SalePrice'], axis=1)
y = home_data_dummies.SalePrice

train_X, val_X, train_y, val_y = train_test_split(X, y, random_state=1)

# Define hyperparameters to explore
n_estimators_list = [50, 100, 150]
max_depth_list = [None, 10, 20]
min_samples_split_list = [2, 5, 10]

best_mae = float('inf')
best_hyperparameters = {}

for n_estimators in n_estimators_list:
    for max_depth in max_depth_list:
        for min_samples_split in min_samples_split_list:
            rf_model = RandomForestRegressor(n_estimators=n_estimators, max_depth=max_depth, min_samples_split=min_samples_split, random_state=1, max_leaf_nodes=300)
            rf_model.fit(train_X, train_y)
            rf_val_predictions = rf_model.predict(val_X)
            rf_val_mae = mean_absolute_error(rf_val_predictions, val_y)
            
            print("n_estimators: {}, max_depth: {}, min_samples_split: {} \t\t Mean Absolute Error: {}".format(n_estimators, max_depth, min_samples_split, rf_val_mae))
            
            if rf_val_mae < best_mae:
                best_mae = rf_val_mae
                best_hyperparameters = {'n_estimators': n_estimators, 'max_depth': max_depth, 'min_samples_split': min_samples_split}

print("\nBest Hyperparameters:")
print(best_hyperparameters)
print("Best Mean Absolute Error:", best_mae)"""

'X = home_data_dummies[list(high_corr_features_unique)]\n#X = home_data_dummies.drop([\'SalePrice\'], axis=1)\ny = home_data_dummies.SalePrice\n\ntrain_X, val_X, train_y, val_y = train_test_split(X, y, random_state=1)\n\n# Define hyperparameters to explore\nn_estimators_list = [50, 100, 150]\nmax_depth_list = [None, 10, 20]\nmin_samples_split_list = [2, 5, 10]\n\nbest_mae = float(\'inf\')\nbest_hyperparameters = {}\n\nfor n_estimators in n_estimators_list:\n    for max_depth in max_depth_list:\n        for min_samples_split in min_samples_split_list:\n            rf_model = RandomForestRegressor(n_estimators=n_estimators, max_depth=max_depth, min_samples_split=min_samples_split, random_state=1, max_leaf_nodes=300)\n            rf_model.fit(train_X, train_y)\n            rf_val_predictions = rf_model.predict(val_X)\n            rf_val_mae = mean_absolute_error(rf_val_predictions, val_y)\n            \n            print("n_estimators: {}, max_depth: {}, min_samples_split: {} \t\t Mean Ab

# Train in all data with best hyperparameters

In [34]:
high_corr_features_unique_test = set(test_data_dummies.columns).intersection(high_corr_features_unique)

In [35]:
common_columns = list(set(home_data_dummies[list(high_corr_features_unique)].columns).intersection(test_data_dummies[list(high_corr_features_unique_test)].columns))

In [36]:
X = X = home_data_dummies[common_columns] #already without SalePrice
y = home_data_dummies.SalePrice

# To improve accuracy, train on all training data
rf_model_on_full_data = RandomForestRegressor(random_state=1, max_leaf_nodes=300, n_estimators= 150, max_depth= 20, min_samples_split= 2)
rf_model_on_full_data.fit(X, y)

RandomForestRegressor(max_depth=20, max_leaf_nodes=300, n_estimators=150,
                      random_state=1)

In [37]:
# make predictions which we will submit. 
test_preds_modelV2 = rf_model_on_full_data.predict(test_data_dummies[common_columns])

test_preds_modelV2[0:15].round(0).tolist()

[128848.0,
 153590.0,
 179925.0,
 183635.0,
 197032.0,
 181985.0,
 169263.0,
 175092.0,
 182897.0,
 119528.0,
 195737.0,
 95207.0,
 101606.0,
 155503.0,
 141426.0]

# Generate a submission

Run the code cell below to generate a CSV file with your predictions that you can use to submit to the competition.

In [38]:
# Run the code to save predictions in the format used for competition scoring

output = pd.DataFrame({'Id': test_data.index,
                       'SalePrice': test_preds_modelV2})
output.to_csv('submission_v4.csv', index=False)

# Submit to the competition

To test your results, you'll need to join the competition (if you haven't already).  So open a new window by clicking on **[this link](https://www.kaggle.com/c/home-data-for-ml-course)**.  Then click on the **Join Competition** button.

![join competition image](https://storage.googleapis.com/kaggle-media/learn/images/axBzctl.png)

Next, follow the instructions below:
1. Begin by clicking on the **Save Version** button in the top right corner of the window.  This will generate a pop-up window.  
2. Ensure that the **Save and Run All** option is selected, and then click on the **Save** button.
3. This generates a window in the bottom left corner of the notebook.  After it has finished running, click on the number to the right of the **Save Version** button.  This pulls up a list of versions on the right of the screen.  Click on the ellipsis **(...)** to the right of the most recent version, and select **Open in Viewer**.  This brings you into view mode of the same page. You will need to scroll down to get back to these instructions.
4. Click on the **Data** tab near the top of the screen.  Then, click on the file you would like to submit, and click on the **Submit** button to submit your results to the leaderboard.

You have now successfully submitted to the competition!

If you want to keep working to improve your performance, select the **Edit** button in the top right of the screen. Then you can change your code and repeat the process. There's a lot of room to improve, and you will climb up the leaderboard as you work.


# Continue Your Progress
There are many ways to improve your model, and **experimenting is a great way to learn at this point.**

The best way to improve your model is to add features.  To add more features to the data, revisit the first code cell, and change this line of code to include more column names:
```python
features = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']
```

Some features will cause errors because of issues like missing values or non-numeric data types.  Here is a complete list of potential columns that you might like to use, and that won't throw errors:
- 'MSSubClass'
- 'LotArea'
- 'OverallQual' 
- 'OverallCond' 
- 'YearBuilt'
- 'YearRemodAdd' 
- '1stFlrSF'
- '2ndFlrSF' 
- 'LowQualFinSF' 
- 'GrLivArea'
- 'FullBath'
- 'HalfBath'
- 'BedroomAbvGr' 
- 'KitchenAbvGr' 
- 'TotRmsAbvGrd' 
- 'Fireplaces' 
- 'WoodDeckSF' 
- 'OpenPorchSF'
- 'EnclosedPorch' 
- '3SsnPorch' 
- 'ScreenPorch' 
- 'PoolArea' 
- 'MiscVal' 
- 'MoSold' 
- 'YrSold'

Look at the list of columns and think about what might affect home prices.  To learn more about each of these features, take a look at the data description on the **[competition page](https://www.kaggle.com/c/home-data-for-ml-course/data)**.

After updating the code cell above that defines the features, re-run all of the code cells to evaluate the model and generate a new submission file.  


# What's next?

As mentioned above, some of the features will throw an error if you try to use them to train your model.  The **[Intermediate Machine Learning](https://www.kaggle.com/learn/intermediate-machine-learning)** course will teach you how to handle these types of features. You will also learn to use **xgboost**, a technique giving even better accuracy than Random Forest.

The **[Pandas](https://kaggle.com/Learn/Pandas)** course will give you the data manipulation skills to quickly go from conceptual idea to implementation in your data science projects. 

You are also ready for the **[Deep Learning](https://kaggle.com/Learn/intro-to-Deep-Learning)** course, where you will build models with better-than-human level performance at computer vision tasks.

---




*Have questions or comments? Visit the [course discussion forum](https://www.kaggle.com/learn/intro-to-machine-learning/discussion) to chat with other learners.*